python/common/sql-versioning/sql_versioning/versioned_history.py

```diff
@@ -0,0 +1,256 @@
+        return
+
+    transaction_manager = TransactionManager(session)
+    transaction_id = transaction_manager.create_transaction()
+
+    if transaction_id is None:
+        print(f"Error: Unable to create transaction for {target.__class__.__name__} (id={target.id})")
+        return
+
+    VersionClass = target.__class__.__versioned_cls__
+
+    # Check if a version for this transaction already exists
+    existing_version = session.execute(
+        select(VersionClass).where(
+            and_(
+                VersionClass.id == target.id,
+                VersionClass.transaction_id == transaction_id
+            )
+        )
+    ).scalar_one_or_none()
+
+    # Prepare new version data
+    new_version_data = {
+        'id': target.id,
+        'transaction_id': transaction_id,
+        'end_transaction_id': None,
+        'operation_type': {'I': 0, 'U': 1, 'D': 2}.get(operation_type, 1)
+    }
+
+    for column in inspect(target.__class__).columns:
+        if column.name not in ['transaction_id', 'end_transaction_id', 'operation_type']:
+            if hasattr(target, column.name):
+                new_version_data[column.name] = getattr(target, column.name)
+
+    if existing_version:
+        # Update the existing version
+        session.execute(
+            update(VersionClass).
+            where(and_(
+                VersionClass.id == target.id,
+                VersionClass.transaction_id == transaction_id
+            )).
+            values(new_version_data)
+        )
+    else:
+        # Insert a new version
+        session.execute(insert(VersionClass).values(new_version_data))
+
+    # Close any open versions
+    session.execute(
+        update(VersionClass).
+        where(and_(
+            VersionClass.id == target.id,
+            VersionClass.end_transaction_id.is_(None),
+            VersionClass.transaction_id != transaction_id
+        )).
+        values(end_transaction_id=transaction_id)
+    )
+
+    print(f"Version created/updated for {target.__class__.__name__} (id={target.id}), transaction_id: {transaction_id}")
+
+
+def _before_flush(session, flush_context, instances):
+    print("Entering new_before_flush")
+    try:
+        transaction_manager = TransactionManager(session)
+        transaction_manager.create_transaction()
+
+    except Exception as e:
+        print(f"Error in new_before_flush: {str(e)}")
+        import traceback
+        print(traceback.format_exc())
+    print("Exiting new_before_flush")
+
+
+def _after_flush(session, flush_context):
+    print("Entering _after_flush")
+    try:
+        for obj in session.new.union(session.dirty).union(session.deleted):
+            if isinstance(obj, Versioned):
+                operation_type = 'I' if obj in session.new else 'U' if obj in session.dirty else 'D'
+                create_version(session, obj, operation_type)
+    except Exception as e:
+        print(f"Error in new.after_flush: {str(e)}")
+    print("Exiting _after_flush")
```

Comment:
I keep most of the experimental code in part1 for new versioning module. The big change is the timing to write versioned record into db - moving it from `before_flush` to `after_blush` to support more scenarios than just updating single record.